File: src/services/attendance/attendanceTimeService.ts
// src/services/attendance/attendanceTimeService.ts
// ============================================================================
// WorkForceOS — Centralized Attendance Time & Timezone Service (Web)
// Authoritative UTC Database Contract with Organization Timezone Rendering (Asia/Kolkata)
// ============================================================================

export class AttendanceTimeService {
  public static readonly DEFAULT_TIMEZONE = 'Asia/Kolkata';
  public static readonly IST_OFFSET_MINUTES = 330; // +05:30 (5 hours 30 minutes)

  /**
   * Resolves timezone offset in minutes
   */
  public static getTimezoneOffsetMinutes(timezone: string = AttendanceTimeService.DEFAULT_TIMEZONE): number {
    const tz = timezone.toLowerCase();
    if (tz.includes('kolkata') || tz.includes('ist') || tz.includes('india') || tz.includes('calcutta')) {
      return AttendanceTimeService.IST_OFFSET_MINUTES;
    }
    return AttendanceTimeService.IST_OFFSET_MINUTES; // Default to India Standard Time
  }

  /**
   * Robustly parses any backend raw string/Date into a valid UTC Date
   */
  public static parseToDate(raw: string | Date | undefined): Date | null {
    if (!raw) return null;
    if (raw instanceof Date) return isNaN(raw.getTime()) ? null : raw;
    const str = String(raw).trim();
    if (!str || str === 'null' || str === '—') return null;

    // 1. If it contains a date (YYYY-MM-DD...)
    if (/^\d{4}-\d{2}-\d{2}/.test(str)) {
      let iso = str.replace(' ', 'T');
      // If Postgres timestamptz like "2026-08-25 12:34:32+00"
      if (!iso.endsWith('Z') && !iso.includes('+') && !iso.includes('-0') && !iso.includes('-1')) {
        iso = `${iso}Z`;
      }
      const d = new Date(iso);
      return isNaN(d.getTime()) ? null : d;
    }

    // 2. Plain time strings: 12-hour format (e.g. "12:34 PM", "03:12:41 PM")
    const match12 = str.match(/^(\d{1,2}):(\d{2})(?::(\d{2}))?\s*(AM|PM)$/i);
    if (match12) {
      let h = parseInt(match12[1], 10);
      const m = parseInt(match12[2], 10);
      const s = match12[3] ? parseInt(match12[3], 10) : 0;
      const isPm = match12[4].toUpperCase() === 'PM';
      if (isPm && h < 12) h += 12;
      if (!isPm && h === 12) h = 0;
      const now = new Date();
      // Assume stored time from punch event is in UTC
      return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), h, m, s));
    }

    // 3. Plain time strings: 24-hour format (e.g. "12:34:32", "18:04")
    const match24 = str.match(/^(\d{1,2}):(\d{2})(?::(\d{2}))?$/);
    if (match24) {
      const h = parseInt(match24[1], 10);
      const m = parseInt(match24[2], 10);
      const s = match24[3] ? parseInt(match24[3], 10) : 0;
      const now = new Date();
      return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), h, m, s));
    }

    const fallback = new Date(str);
    return isNaN(fallback.getTime()) ? null : fallback;
  }

  /**
   * Converts any Date or UTC string to the organization local Date
   */
  public static toOrganizationTime(raw: string | Date | undefined, timezone: string = AttendanceTimeService.DEFAULT_TIMEZONE): Date | null {
    const d = AttendanceTimeService.parseToDate(raw);
    if (!d) return null;
    const offsetMs = AttendanceTimeService.getTimezoneOffsetMinutes(timezone) * 60000;
    return new Date(d.getTime() + offsetMs);
  }

  /**
   * Determines the authoritative business attendance date (YYYY-MM-DD) in the organization timezone
   */
  public static getOrganizationBusinessDate(date: Date = new Date(), timezone: string = AttendanceTimeService.DEFAULT_TIMEZONE): string {
    const local = AttendanceTimeService.toOrganizationTime(date, timezone);
    if (!local) {
      const now = new Date();
      return `${now.getUTCFullYear()}-${String(now.getUTCMonth() + 1).padStart(2, '0')}-${String(now.getUTCDate()).padStart(2, '0')}`;
    }
    return `${local.getUTCFullYear()}-${String(local.getUTCMonth() + 1).padStart(2, '0')}-${String(local.getUTCDate()).padStart(2, '0')}`;
  }

  /**
   * Formats any backend timestamptz or raw UTC time into authoritative 12-hour display string in Asia/Kolkata (IST)
   * e.g. "10:09:47 AM" or "06:04:32 PM"
   */
  public static formatAttendanceTime(
    raw: string | Date | undefined,
    timezone: string = AttendanceTimeService.DEFAULT_TIMEZONE,
    withSeconds: boolean = false,
    fallback: string = '—'
  ): string {
    if (!raw) return fallback;
    const str = String(raw).trim();
    if (!str || str === 'null' || str === '—' || str === '--:--' || str === '-') return fallback;

    // 1. If already 12-hour AM/PM string (e.g. "10:09 AM", "03:12:41 PM")
    const match12 = str.match(/^(\d{1,2}):(\d{2})(?::(\d{2}))?\s*(AM|PM)$/i);
    if (match12) {
      const h = match12[1].padStart(2, '0');
      const m = match12[2];
      const s = match12[3];
      const ampm = match12[4].toUpperCase();
      return withSeconds && s ? `${h}:${m}:${s} ${ampm}` : `${h}:${m} ${ampm}`;
    }

    // 2. If plain 24-hour time string without date (e.g. "10:09:47", "18:04")
    const match24 = str.match(/^(\d{1,2}):(\d{2})(?::(\d{2}))?$/);
    if (match24) {
      let h = parseInt(match24[1], 10);
      const m = match24[2];
      const s = match24[3] || '00';
      const suffix = h >= 12 ? 'PM' : 'AM';
      h = h % 12 === 0 ? 12 : h % 12;
      const hh = String(h).padStart(2, '0');
      return withSeconds ? `${hh}:${m}:${s} ${suffix}` : `${hh}:${m} ${suffix}`;
    }

    // 3. ISO DateTime string or Postgres timestamptz (e.g. "2026-08-26T04:39:47.123Z" or "2026-08-26 04:39:47+00")
    try {
      const iso = str.includes('T') ? str : str.replace(' ', 'T');
      const normalized = (!iso.endsWith('Z') && !iso.includes('+') && !iso.includes('-0') && !iso.includes('-1'))
        ? `${iso}Z`
        : iso;
      const date = new Date(normalized);
      if (!isNaN(date.getTime())) {
        return new Intl.DateTimeFormat('en-US', {
          timeZone: timezone || AttendanceTimeService.DEFAULT_TIMEZONE,
          hour: '2-digit',
          minute: '2-digit',
          second: withSeconds ? '2-digit' : undefined,
          hour12: true,
        }).format(date);
      }
    } catch (_) {}

    return str;
  }

  /**
   * Calculates gross and net working minutes, late minutes, and overtime
   */
  public static calculateWorkingMinutes(
    firstInRaw?: string | Date,
    lastOutRaw?: string | Date,
    expectedIn: string = '09:00 AM',
    expectedOut: string = '06:00 PM',
    breakMinutes: number = 45
  ): {
    grossMinutes: number;
    netMinutes: number;
    lateMinutes: number;
    earlyMinutes: number;
    overtimeMinutes: number;
  } {
    if (!firstInRaw) {
      return { grossMinutes: 0, netMinutes: 0, lateMinutes: 0, earlyMinutes: 0, overtimeMinutes: 0 };
    }

    const inDate = AttendanceTimeService.toOrganizationTime(firstInRaw);
    if (!inDate) {
      return { grossMinutes: 0, netMinutes: 0, lateMinutes: 0, earlyMinutes: 0, overtimeMinutes: 0 };
    }

    const outDate = lastOutRaw ? AttendanceTimeService.toOrganizationTime(lastOutRaw) : null;
    let grossMinutes = 0;

    if (outDate && outDate > inDate) {
      grossMinutes = Math.floor((outDate.getTime() - inDate.getTime()) / 60000);
    }

    const netMinutes = Math.max(0, grossMinutes - breakMinutes);

    // Calculate late arrival
    const inTotalMins = inDate.getHours() * 60 + inDate.getMinutes();
    const expInMatch = expectedIn.match(/(\d+):(\d+)\s*(AM|PM)?/i);
    let expInMins = 540; // 09:00 AM default
    if (expInMatch) {
      let eh = parseInt(expInMatch[1], 10);
      const em = parseInt(expInMatch[2], 10);
      if (expInMatch[3]?.toUpperCase() === 'PM' && eh < 12) eh += 12;
      if (expInMatch[3]?.toUpperCase() === 'AM' && eh === 12) eh = 0;
      expInMins = eh * 60 + em;
    }
    const lateMinutes = Math.max(0, inTotalMins - expInMins);

    // Overtime
    const overtimeMinutes = Math.max(0, netMinutes - 480);

    return {
      grossMinutes,
      netMinutes,
      lateMinutes,
      earlyMinutes: 0,
      overtimeMinutes,
    };
  }

  /**
   * Diagnostic logger
   */
  public static logDiagnostic(context: string, raw: any, timezone: string = AttendanceTimeService.DEFAULT_TIMEZONE): void {
    const formatted = AttendanceTimeService.formatAttendanceTime(raw, timezone, true);
    console.log(`[ATTENDANCE_TIME_DIAGNOSTIC][${context}] RAW:`, raw, `| TZ: ${timezone} | DISPLAY: ${formatted}`);
  }
}

export const attendanceTimeService = AttendanceTimeService;

File: 1.FlutterApp/flutter_app/lib/core/services/attendance_time_service.dart
// attendance_time_service.dart
// ============================================================================
// WorkForceOS — Centralized Attendance Time & Timezone Service
// Enforces UTC Database Contract with Authoritative Organization Timezone Rendering
// ============================================================================

import '../utils/secure_log.dart';

class AttendanceTimeService {
  static const String defaultTimezone = 'Asia/Kolkata';
  static const Duration istOffset = Duration(hours: 5, minutes: 30);

  /// Resolves the timezone offset for a configured timezone identifier
  static Duration getOffset([String? timezone]) {
    final tz = (timezone ?? defaultTimezone).toLowerCase();
    if (tz.contains('kolkata') || tz.contains('ist') || tz.contains('india') || tz.contains('calcutta')) {
      return istOffset;
    }
    return istOffset; // Default to India Standard Time (UTC+05:30)
  }

  /// Converts any [DateTime] instance to the authoritative organization local time
  static DateTime toOrganizationTime(DateTime dt, [String? timezone]) {
    final utc = dt.toUtc();
    final offset = getOffset(timezone);
    return utc.add(offset);
  }

  /// Parses a raw backend timestamptz string or ISO timestamp into an organization local [DateTime].
  ///
  /// Supports:
  ///   - Postgres timestamptz: '2026-08-25T09:42:41.780231+00:00', '2026-08-25 12:34:32.174602+00'
  ///   - ISO 8601 UTC: '2026-08-25T09:42:41Z'
  ///   - Plain TIME strings: '15:12:41', '09:30 AM', '18:30'
  static DateTime? parseServerTimestamp(dynamic raw, {String? timezone, String? referenceDate}) {
    if (raw == null) return null;
    final str = raw.toString().trim();
    if (str.isEmpty || str == 'null' || str == '—') return null;

    // 1. Check for ISO / Postgres timestamptz with date component
    final parsed = DateTime.tryParse(str);
    if (parsed != null) {
      return toOrganizationTime(parsed, timezone);
    }

    // 2. Parse 12-hour or 24-hour time strings (e.g. "03:12 PM", "15:12:41", "09:30")
    final match12 = RegExp(r'^(\d{1,2}):(\d{2})(?::(\d{2}))?\s*(AM|PM)$', caseSensitive: false).firstMatch(str);
    if (match12 != null) {
      int h = int.parse(match12.group(1)!);
      final m = int.parse(match12.group(2)!);
      final s = match12.group(3) != null ? int.parse(match12.group(3)!) : 0;
      final isPm = match12.group(4)!.toUpperCase() == 'PM';
      if (isPm && h < 12) h += 12;
      if (!isPm && h == 12) h = 0;

      final base = _resolveBaseDate(referenceDate, timezone);
      return DateTime(base.year, base.month, base.day, h, m, s);
    }

    final match24 = RegExp(r'^(\d{1,2}):(\d{2})(?::(\d{2}))?').firstMatch(str);
    if (match24 != null) {
      final h = int.parse(match24.group(1)!);
      final m = int.parse(match24.group(2)!);
      final s = match24.group(3) != null ? int.parse(match24.group(3)!) : 0;

      final base = _resolveBaseDate(referenceDate, timezone);
      return DateTime(base.year, base.month, base.day, h, m, s);
    }

    return null;
  }

  static DateTime _resolveBaseDate(String? referenceDate, String? timezone) {
    if (referenceDate != null && referenceDate.isNotEmpty) {
      final p = DateTime.tryParse(referenceDate);
      if (p != null) return p;
    }
    return toOrganizationTime(DateTime.now().toUtc(), timezone);
  }

  /// Determines the authoritative attendance business date (`YYYY-MM-DD`) in organization timezone.
  /// (Crucial: 20:30 UTC on Aug 24 is 02:00 AM Aug 25 in Asia/Kolkata -> Returns '2026-08-25').
  static String getBusinessDate([DateTime? utcInstant, String? timezone]) {
    final instant = utcInstant ?? DateTime.now().toUtc();
    final local = toOrganizationTime(instant, timezone);
    return '${local.year}-${local.month.toString().padLeft(2, '0')}-${local.day.toString().padLeft(2, '0')}';
  }

  /// Formats any backend timestamp or time string into a clean display time (e.g. "03:12:41 PM" or "03:12 PM").
  static String formatAttendanceTime(
    dynamic raw, {
    bool withSeconds = false,
    String? timezone,
    String? referenceDate,
    String fallback = '—',
  }) {
    if (raw == null) return fallback;
    final str = raw.toString().trim();
    if (str.isEmpty || str == 'null' || str == '—') return fallback;

    final dt = parseServerTimestamp(raw, timezone: timezone, referenceDate: referenceDate);
    if (dt == null) return raw.toString();

    final h24 = dt.hour;
    final m = dt.minute;
    final s = dt.second;
    final suffix = h24 >= 12 ? 'PM' : 'AM';
    final h12 = h24 % 12 == 0 ? 12 : h24 % 12;

    final hh = h12.toString().padLeft(2, '0');
    final mm = m.toString().padLeft(2, '0');

    if (withSeconds) {
      final ss = s.toString().padLeft(2, '0');
      return '$hh:$mm:$ss $suffix';
    }
    return '$hh:$mm $suffix';
  }

  /// Formats attendance business date into "25 Aug, 2026" or "Tuesday, 25 Aug 2026"
  static String formatAttendanceDate(dynamic raw, {bool full = false, String? timezone, String fallback = '—'}) {
    if (raw == null) return fallback;
    final str = raw.toString().trim();
    if (str.isEmpty || str == 'null' || str == '—') return fallback;

    DateTime? dt = DateTime.tryParse(str);
    dt ??= parseServerTimestamp(raw, timezone: timezone);
    if (dt == null) return str;

    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];

    final dayName = days[dt.weekday - 1];
    final monthName = months[dt.month - 1];

    if (full) {
      return '$dayName, ${dt.day} $monthName ${dt.year}';
    }
    return '${dt.day} $monthName, ${dt.year}';
  }

  /// Calculates net worked duration between IN and OUT times, accounting for approved breaks
  static Duration calculateDuration({
    DateTime? checkIn,
    DateTime? checkOut,
    int breakMinutes = 0,
    String? timezone,
  }) {
    if (checkIn == null) return Duration.zero;
    final end = checkOut ?? toOrganizationTime(DateTime.now().toUtc(), timezone);
    final totalDiff = end.difference(checkIn);
    if (totalDiff.isNegative) return Duration.zero;

    final netMinutes = totalDiff.inMinutes - breakMinutes;
    return netMinutes > 0 ? Duration(minutes: netMinutes) : Duration.zero;
  }

  /// Formats Duration to HH:MM:SS
  static String formatDurationTimer(Duration d) {
    final h = d.inHours.toString().padLeft(2, '0');
    final m = (d.inMinutes % 60).toString().padLeft(2, '0');
    final s = (d.inSeconds % 60).toString().padLeft(2, '0');
    return '$h:$m:$s';
  }

  /// Formats Duration to readable string e.g. "7h 45m"
  static String formatDurationReadable(Duration d) {
    final h = d.inHours;
    final m = (d.inMinutes % 60).toString().padLeft(2, '0');
    return '${h}h ${m}m';
  }

  /// Prints unified timezone diagnostic log for development debugging
  static void logDiagnostic({
    required String context,
    required dynamic rawValue,
    String? timezone,
  }) {
    final orgTz = timezone ?? defaultTimezone;
    final parsed = parseServerTimestamp(rawValue, timezone: orgTz);
    final display = formatAttendanceTime(rawValue, timezone: orgTz, withSeconds: true);

    secureLog(
      '[ATTENDANCE_TIME_DIAGNOSTIC][$context] '
      'RAW: $rawValue | ORG_TZ: $orgTz | LOCAL: $parsed | DISPLAY: $display',
    );
  }
}
